NIL 129 tabbed search

mchalejeiwc opened the pull request:
This is ready for formal review now.

This PR provides tabbed search using 'scope' to determine which content will be surfaced. Scope derivation is now more elegant than the first iteration.

Tab structure/style follows the Gov site as has been agreed.

Note as there is no (as yet) 'services' or 'news' content, the acceptance test asserts this.

jacsamell commented on repository-data/webfiles/src/main/resources/site/freemarker/common/searchtabs.ftl:
```diff
@@ -0,0 +1,13 @@
+    
+    <ul class="tabs-nav" role="tablist">
+        <#list tabs as tab>    
+            <li class="${(area=='${tab}')?string("active", "")}"><a tabindex="0" href="${tabLink}${tab}" title="<@fmt.message key="searchtab.${tab}"/>"><@fmt.message key="searchtab.${tab}"/></a></li>
```
What does the tabindex=0 mean? Should this not increment for the tabs?

mchalejeiwc replied:
Should be tab?index, will fix

ziemsky commented on acceptance-tests/src/test/resources/features/site/search.feature:
```diff
@@ -145,6 +145,25 @@ Feature: Basic search
             | Series / Collection ( ... |
             | Methodology ( ...         |
 
+    Scenario: Clicking on the 'Data and Information' tab displays the results page with full, unfiltered result set
```
As discussed, this suggests that there is no 'All' tab anymore.

jacsamell commented on site/src/main/java/uk/nhs/digital/common/components/SearchComponent.java:
```diff
@@ -192,15 +204,46 @@ private HstQuery buildQuery(HstRequest request) {
+
+        List<HippoBean> scopeBeans = new ArrayList<HippoBean>();
+
+        if (areaParam.equals(SearchArea.NEWS)) {
```
If you're using ENUMs you might want to use a switch here instead, more explicit.

ziemsky replied:
Agreed. Granted, that's replacing one kind of evil with another but the switch/case will be a bit easier to read.

mchalejeiwc replied:
I'm changing to use a switch ;)

jacsamell commented on site/src/main/java/uk/nhs/digital/common/components/SearchComponent.java:
```diff
@@ -192,15 +204,46 @@ private HstQuery buildQuery(HstRequest request) {
     }
 
+    private SearchArea getAreaOption(HstRequest request) {
+        return SearchArea.valueOf(Optional.ofNullable(getAnyParameter(request, REQUEST_PARAM_AREA)).orElse(AREA_DEFAULT).toUpperCase());
```
I'd use AREA_DEFAULT as the enum itself rather than converting it to a string and then back to an enum.

```
String area = getAnyParameter(request, REQUEST_PARAM_AREA)

return area == null ? AREA_DEFAULT : SearchArea.valueOf(area)
```

mchalejeiwc replied:
Good call

ziemsky commented on repository-data/webfiles/src/main/resources/site/freemarker/common/base-layout.ftl:
```diff
@@ -88,11 +88,22 @@
           </div>
         </div><!--
         --><div class="layout__item layout-2-3">
+          <#if hstResponseChildContentNames?seq_contains("tabs") >
```
Small that this `if` block is, I think it warrants turning into a macro, which could then be applied here (lines 91-95) and in lines 102-106.

mchalejeiwc replied:
Will do

mchalejeiwc commented:
Updated to use Switch over If/Else, extracted macro for the search bar, and enabled the 'feature toggle'. This defaults to 'true' for local and acceptance tests, so I'll need to pre-populate that to 'false' in PROD.
